numbat/src/typechecker/map_stack.rs

```diff
@@ -0,0 +1,80 @@
+use std::hash::Hash;
+use std::{borrow::Borrow, collections::HashMap};
+
+/// A stack of hash maps. All insertions affect the top hash map, preserving
+/// any entries in maps below. The `save` function can be used to push a new
+/// map on the top of the stack, in effect saving the current state of the map,
+/// which one can then restore with `restore`.
+#[derive(Debug, Clone)]
+pub(crate) struct MapStack<K, V> {
+    top: HashMap<K, V>,
+    stack: Vec<HashMap<K, V>>,
+}
+
+impl<K, V> Default for MapStack<K, V> {
+    fn default() -> Self {
+        MapStack {
+            top: Default::default(),
+            stack: Default::default(),
+        }
+    }
+}
+
+impl<K: Hash + Eq, V> MapStack<K, V> {
+    fn iter_dict(&self) -> impl Iterator<Item = &HashMap<K, V>> {
+        std::iter::once(&self.top).chain(self.stack.iter())
+    }
+
+    fn iter_dict_mut(&mut self) -> impl Iterator<Item = &mut HashMap<K, V>> {
+        std::iter::once(&mut self.top).chain(self.stack.iter_mut())
+    }
+
+    pub(crate) fn iter(&self) -> impl Iterator<Item = (&K, &V)> {
+        self.iter_dict().flatten()
+    }
+
+    pub(crate) fn iter_mut(&mut self) -> impl Iterator<Item = (&K, &mut V)> {
+        self.iter_dict_mut().flatten()
+    }
+
+    pub(crate) fn keys(&self) -> impl Iterator<Item = &K> {
+        self.iter_dict().map(|dict| dict.keys()).flatten()
+    }
+
+    pub(crate) fn get<Q>(&self, key: &Q) -> Option<&V>
+    where
+        K: Borrow<Q>,
+        Q: Hash + Eq + ?Sized,
+    {
+        self.iter_dict().filter_map(|dict| dict.get(key)).nth(0)
+    }
+
+    pub(crate) fn insert(&mut self, key: K, value: V) -> Option<V> {
+        self.top.insert(key, value)
```

Comment:
Does this yield proper semantics? `HashMap::insert` returns `Some(…)` if the map previously contained the value that is being inserted. But you are not checking the other maps on the stack here.

It might not be problematic for our current usage of the map (maybe we never check the return value of `insert`?). But then we should maybe change the return type to `()`?

Reply:
I've changed it to `()`, it makes more sense indeed